include/librpmi.h

```diff
@@ -783,6 +796,230 @@ struct rpmi_service_group {
+				       rpmi_uint64_t *rate);
+
+	/**
+	 * Check if the requested rate is not in the allowed margin(Hz)
+	 * which require change in clock rate.
+	 **/
+	rpmi_bool_t (*rate_change_match)(void *priv,
+				  rpmi_uint32_t clock_id,
+				  rpmi_uint64_t rate);
+
+	/**
+	 * Set clock rate.
+	 * Also based on the rate match mode and PLL lock frequency
+	 * the actual frequency set may have +-margin with requested rate.
+	 * Return the set rate in new_rate buffer
+	 * */
+	enum rpmi_error (*set_rate)(void *priv,
+				    rpmi_uint32_t clock_id,
+				    enum rpmi_clock_rate_match match,
+				    rpmi_uint64_t rate,
+				    rpmi_uint64_t *new_rate);
+
+	/**
+	 * Recalculate and set rate.
+	 * Recalculate and set the clock rate based on the new input(parent)
+	 * clock and return the new rate in buffer.
+	 */
+	enum rpmi_error (*set_rate_recalc)(void *priv,
+					rpmi_uint32_t clock_id,
+					rpmi_uint64_t parent_rate,
+					rpmi_uint64_t *new_rate);
+};
+
+/**
+ * @brief Get number of clocks
+ *
+ * @param[in] clock_mod		pointer to clock module instance
+ * @return rpmi_uint32_t	number of clocks
+ */
+rpmi_uint32_t rpmi_clock_get_clock_count(struct rpmi_clock_module *clock_mod);
+
+/**
+ * @brief Get clock attributes
+ *
+ * @param[in] clock_mod		pointer to clock_module instance
+ * @param[in] clock_id		clock id
+ * @param[in] attrs		pointer to attribute buffer
+ * @return enum rpmi_error
+ */
+enum rpmi_error rpmi_clock_get_attrs(struct rpmi_clock_module *clock_mod,
+				     rpmi_uint32_t clock_id,
+				     struct rpmi_clock_attrs *attrs);
+
+/**
+ * @brief Set clock state
+ *
+ * @param[in] clock_mod		pointer to clock_module instance
+ * @param[in] clock_id		clock id
+ * @param[in] state		new clock state
+ * @return enum rpmi_error
+ */
+enum rpmi_error rpmi_clock_set_state(struct rpmi_clock_module *clock_mod,
+				     rpmi_uint32_t clock_id,
+				     enum rpmi_clock_state state);
+
+/**
+ * @brief Get clock current state
+ *
+ * @param[in] clock_mod		pointer to clock_module instance
+ * @param[in] clock_id		clock id
+ * @param[in] state		pointer to state buffer
+ * @return enum rpmi_error
+ */
+enum rpmi_error rpmi_clock_get_state(struct rpmi_clock_module *clock_mod,
+				     rpmi_uint32_t clock_id,
+				     enum rpmi_clock_state *state);
+
+/**
+ * @brief Set clock rate
+ *
+ * @param[in] clock_mod		pointer to clock_module instance
+ * @param[in] clock_id		clock id
+ * @param[in] match		clock rate match mode
+ * @param[in] rate		new clock rate in Hertz
+ * @return enum rpmi_error
+ */
+enum rpmi_error rpmi_clock_set_rate(struct rpmi_clock_module *clock_mod,
+				    rpmi_uint32_t clock_id,
+				    enum rpmi_clock_rate_match match,
+				    rpmi_uint64_t rate);
+
+/**
+ * @brief Get clock rate
+ *
+ * @param[in] clock_mod		pointer to clock_module instance
+ * @param[in] clock_id		clock id
+ * @param[in] rate		pointer to rate buffer
+ * @return enum rpmi_error
+ */
+enum rpmi_error rpmi_clock_get_rate(struct rpmi_clock_module *clock_mod,
```

Comment:
I don't understand why we need to expose rpmi_clock_get_clock_count(), rpmi_clock_get_attrs(), rpmi_clock_set_state(), rpmi_clock_set_rate(), and rpmi_clock_get_rate() as public APIs.

These are supposed to be internal functions of librpmi.

Reply:
I will merge the rpmi_clock.c with *service_group_clock.c. so these will go away from header

Reply:
I made them separate to provide interface apart from service group directly to host fw in case. Since we are unsure of all the usecases currently

Reply:
In the future, we can selectively expose APIs in librpmi based on use-case.